File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cs
public class Solution {
    public int LengthOfLongestSubstring(string s) {
        HashSet<char> hs = new HashSet<char>();
        int l = 0;
        int longest = 0;
        int res = 0;

        for (int i = 0; i < s.Length; i++)
        {
            char letter = s[i];
            while (hs.Contains(letter))
            {
                hs.Remove(s[l]);
                l += 1;
            }

            hs.Add(letter);
            if (i - l + 1 > res)
                res = i - l + 1;
        }

        return res;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        d = {}
        res = 0

        l = 0
        for r in range(len(s)):
            letter = s[r]
            d[letter] = d.get(letter, 0) + 1
            while d[letter] > 1:
                d[s[l]] -= 1
                l += 1

            res = max(res, r - l + 1)
        
        return res